client/extensions/woocommerce/state/ui/products/edits-reducer.js

```diff
@@ -0,0 +1,59 @@
+/**
+ * External dependencies
+ */
+import { isNumber } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import {
+	WOOCOMMERCE_EDIT_PRODUCT,
+} from '../../action-types';
+
+const initialState = null;
+
+export default function( state = initialState, action ) {
+	const handlers = {
+		[ WOOCOMMERCE_EDIT_PRODUCT ]: editProductAction,
+	};
+
+	const handler = handlers[ action.type ];
+
+	return ( handler && handler( state, action ) ) || state;
+}
+
+function editProductAction( edits, action ) {
+	const { product, data } = action.payload;
+
+	const prevEdits = edits || {};
+	const bucket = product && isNumber( product.id ) && 'updates' || 'creates';
+	const array = editProduct( prevEdits[ bucket ], product, data );
+
+	return { ...prevEdits, [ bucket ]: array };
+}
+
+function editProduct( array, product, data ) {
+	// Use the existing product id (real or placeholder), or creates.length if no product.
+	const prevArray = array || [];
+	const productId = ( product ? product.id : { index: prevArray.length } );
```

Comment:
Maybe makes sense to have the same check as above?
```js
isNumber( product.id ) ? product.id : { index: prevArray.length }
```
Would product ever be defined as `{ }` in the payload of the event?